doc/design/block_execution.md

```diff
@@ -0,0 +1,147 @@
+
+- forward pass: generate OpDesc and VarDesc for forward operators
+
+- backward pass: generate OpDesc and VarDesc for backward operators
+
+- placement pass: generate OpDesc and VarDesc for send/recv/copy operators. Data Parallelism and Model Parallelism will be considered at this stage. We have a placement policy to set concrete device id for every operator. And send/recv/copy operators will be inserted at correct position. 
+
+- optimize pass: generate OpDesc and VarDesc for optimize operators
+
+Each pass will modify the global ProgramDesc.
+
+### Run-time analysis
+
+At run-time, the ProgramDesc generated at compile-time is actually executed on a specific hardware environment. `Executor` is introduced to execute a ProgramDesc. And `Executor` will have a `Run` interface.
+
+In PaddlePaddle, we use DeviceContext to manage hardware resources. There are CPUDeviceContext and CUDADeviceContext for CPU and GPU respectively.A CUDADeviceContext actually is associated with a specific CUDA stream. And Operator will run on a specfic DeviceContext. 
+
+However, users' configuration of hardware is simple, maybe only some GPU card ids will be given. In placement pass of compile-time, only source device id and destination device id will bet set.
+
+So, `Executor` need to have a `DeviceContextManager` to initialize some DeviceContext at the very beginning. `DeviceContextManager` maneges all DeviceContexts in all devices.
+
+The `Executor` is defined as follows:
+
+```
+class Executor {
+public:
+  Executor(ProgramDesc*);
+  void Run();
+  
+private:
+  DeviceContextManager_;
+};
+```
+
+
+## Solution
+
+### Compile-time solution
+  
+#### Remove NetOp
+
+At current code base, we have a NetOp defined as follows:
+
+```
+class NetOp : OperatorBase {
+public:
+  void Run(const Scope&, const platform::DeviceContext&);
+  std::vector<std::unique_ptr<OperatorBase>> ops_;
+};
+```
+
+We can compose some basic operators into a big operator, such [FCOp](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/fc_op.cc). And the backward operators generated by [Backward](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/backward.h) is often a NetOp.
+
+There are mainly two breaking point of NetOp unpon the overall design:
+
+- Backward pass should be do at compile-time, not run-time.
+- A NetOp will corresponds only one OpDesc in BlockDesc. We will have a big backward NetOp and run sequentially. It's hard to do potential memory-resuse/kernel-fusion optimization.
+
+We should generate a group of OpDesc at compile-time instead of a NetOp at run-time.
+
+#### Symbolic API
+
+Please refer to the survey [doc](https://github.com/QiJune/Paddle/blob/e90ec7783a1abe7f7627f97559cc46488e41cc7e/doc/design/graph_survey.md) on Computation Graph. Users will write a neural network topology with Symbolic API. And the composition of operators should be done at compile-time in this level too.
+
+#### Unified Pass Interface
```

Comment:
Need a description of what does "Unified Pass Interface" mean. Maybe can change the title to "Converter" and describe below what does "Converter" mean.

Reply:
Done